### DISCUSSION TITLE
Welcome to JSxCAD Discussions!

### DISCUSSION THREAD

**Discussion by pentacular** [jsxcad/JSxCAD#1155] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by BarbourSmith** [comment]
Hello, I'm Bar.

I'm a big believer that one of the keys to having an ecosystem of open source physical things in the same way that we have an ecosystem of open source software things is a language to build them in.

**Reply by hrgdavor** [reply]
Hi i am Davor Hrg, and I would like to also seemore and more opensource physical things. I love Javascript (but not Typescript) and love doing CAD in JavaScript regardles if it is jscad jsxcad, or WASM of some others like openscad or OCCT.

**Reply by BarbourSmith** [reply]
What are your thoughts on the strengths and weaknesses of the other language options? Is there anything that we should copy from them?

**Reply by hrgdavor** [reply]
If we just take you and me, we will both have different ideas on how to do things :
- some things will be cool things we can learn from each other
- some things will be just different ways to do the same thing

In case of learning from each other it is obvious what to do:  we just exchange ideas. 

In case of doing same thing in a different way it likely just comes to personal preference which way each of us uses. But sharing those can help others to have more options based on if they prefer my or your approach to making something. Whatever API you are creating as part of jsxcad now, some will like, and some might not. 

that said, I am sure there are many things to learn from other language options as there is many more people there.

----------------------
finally to start answering your question:


I would start with one major conceptual difference of CSG vs BREP.  jscad is CSG and I assume jsxcad too.

There are I think nice concepts in BREP to consider if possible adopting .

If you are not familiar yet with CadQuery check it out https://cadquery.readthedocs.io/en/latest/examples.html
it uses similar pattern (builder) like jQuery and I think jsxcad uses similar builder pattern for API

And there is nice JS library around OCCT WASM https://replicad.xyz/

**Reply by pentacular** [reply]
I think that there may be a little confusion on CSG and BREP.

jscad represents solids using a disconnected polygon soup -- this is a boundary representation (BREP).
In contrast CSG uses a composition of volumetric primitives (like sphere, cube, etc).

The only CSG involved in jscad geometry are the half-spaces temporarily constructed within the BSP tree used for boolean operations.
The real involvement of CSG in jscad is that the user api uses similar operations (union, difference, intersection) to CSG systems.

jsxcad/jot (we're changing the name) uses a half-edge face graph to represent solids. This is also a boundary representation.

To be honest BREP vs CSG doesn't matter very much, providing you maintain watertight manifold geometry, which is where jscad really fails, due to its reliance on polygon soup and failure to maintain geometric integrity -- so things will work for a while, but the more operations you use the further the faces will drift out of true, and then everything will collapse when some computation falls through a gap.

jsxcad/jot maintains watertight manifold geometry through volumetric operations, and the face graph makes many operations much simpler (such as finding the geometric outline of a shape, or building a hinge along an edge), but it can be more expensive.

I think the really interesting problem here is about expression -- how can we express complex and useful geometry in code?

I think we've come a fair way toward this, but there's a long way to go yet. :)

**Reply by hrgdavor** [reply]
I am relatively new to the whole technology, so I may be confusing some terms or not understanding completely internals of some of the libraries. 

My main interest is also how to express a model I am trying to create and how to make it faster and more easily to re-iterate on the design.

After the initial hurdle of being able to make some model I imagined, next thing I look to improve is performance. How long it takes from making a change to see the result. If a slow tool is the only one that will allow me to do something I will use it, but will look for alternative all along. For example, I have a way in some of my scripts to turn-off booleans (just show a kind of preview) so I do not wait booleans to finish for each iteration of the design.

I will not use just one tool, but will strongly prefer tools that can be scripted in JavaScript.

Things I am keeping in the possible toolbox
- https://openjscad.xyz/
- https://scriptcad.org/
- https://replicad.xyz/
- https://maker.js.org/
- https://github.com/DSchroer/openscad-wasm

Others as a possible source of inspiration
- https://github.com/CadQuery/cadquery - OCCT
- https://zalo.github.io/CascadeStudio/ - OCCT
- https://brlcad.org/wiki/FAQ
- https://solvespace.com/index.pl

**Reply by pentacular** [reply]
One way to reduce time is to cache results and only recompute those that may have changed.

Turning off booleans is probably not going to work well for cuts or clips, but it's easy to defer joins by grouping instead of joining.

**Reply by hrgdavor** [reply]
Actually turning off booleans worked very well for me.
 - union is trivial (rendering with or without union works great due to how rendering works)
 - subtract (I make base object semi-transparent, and holes solid)

Added benefit of disabling booleans for subtract is that I can see where that cylinder eneded-up when I miss the target object completely while drilling a hole.

preview booleans
![image](https://user-images.githubusercontent.com/2480762/188269785-f7568ba4-55fe-4641-8421-b96dfcc53da5.png)

perform booleans
![image](https://user-images.githubusercontent.com/2480762/188269804-508372f5-4136-4d78-8e25-35b2be146cd1.png)


One big benefit of doing this as a flag allows the code to stay the same, but the rendering to act differently based on the current need.
Also another common performance booster is having different tuneable global precision for converting curves into meshes,
so while iterating design I can lower precision to increase performance and use different precision for final render or export.

Note that example screenshots are not a feature of the library I used, but my own setup inside the script.

**Reply by BarbourSmith** [reply]
Those are some great resources to know about. I'd looked at a couple of them, but I also wasn't aware of others. openscad-wasm seems particularly interesting because it is also a wasm implementation of CGAL so maybe there is some shared value that we could get from them.

On the performance front, have you seen that any of these projects are notably faster or slower than others?

**Reply by hrgdavor** [reply]
I do not know if there are notable performance differences, did not try to compare.

**Reply by pentacular** [reply]
I can see it working in a couple of cases, but not generally, and using transparency and overlays for visualization makes sense.
But this seems a bit fiddly, really.

For this kind of thing I've been using void geometry (geometry representing holes) where disjunction can be deferred until later.
The real win, imho, is in structuring the dependency graph so that you don't need to keep recomputing the expensive operations.

**Reply by hrgdavor** [reply]
Well it is not perfect solution and does not cover all cases. I will use whatever I can at the moment. So I am eager to learn any tricks others use to gain performance while modelling.

I agree with you, modelling library should ideally generate a graph with operations to give more chance for optimizations by recalculating changed bits (jscad does not do this, it executes operations as the code calls each operation). 

General solutions to performance will always fail at the end. I prefer having different reusable tools (or usage patterns) to gain performace when needed, and not relying 100% on the library to be performant whatever crazy thing I do. When I write a script I can decide which parts are more important to cache or disable during development (like the mentioned boolean preview).

When I see a common use case with well defined constraints where I would use general operations, I see as oportunity to make a procedural generator that is optimized. There is one I have not finished for more than a year http://3d.hrg.hr/jscad/three/threejscad2.html?uri=#model.screw.js (I get distracted easily) ... and maybe spend too much time commenting.

Looking forward to learning new stuff from jsxcad and likely using it to model with it too (as one of the tools in the toolbox).

**Reply by pentacular** [reply]
I'm focusing on cleaning up the notebooks and UX at the moment.

If you're interested, you can try it out here:

https://jsxcad.js.org/alpha/#https://raw.githubusercontent.com/jsxcad/JSxCAD/master/nb/start.nb

The UX needs a fair bit of work, and some of the notebooks need maintenance, but feel free to give feedback on things which are annoying or confusing. :)

Please note that it expects chromium, so other browsers may fail in exciting ways.

**Reply by hrgdavor** [reply]
Hi, I took a quick look. I will play with it more to get to know api better in the future.

It likely takes getting used to, the model of working with docs as notebook is a personal preference. You may like it but, I am personally not a fan. too much noise. 
![image](https://user-images.githubusercontent.com/2480762/188391640-639e7527-31c4-4462-b6d4-504129b4d24d.png)

it is hard for me to focus when every md comment is duplicated in this notebook style above

I would personally prefer
![image](https://user-images.githubusercontent.com/2480762/188392407-53f5ea2e-843a-47ea-84c6-9dc7486f32ad.png)


Also font is too large for my taste in the notebook :) 

if you take a look at the code converted to comments below, it looks to me even nicer with github styling for js code blocks

```js
// # Fonts
// Fonts are relatively expensive, so they are imported on demand.

import { readFont, Text } from '@jsxcad/api-v1-font';

// #### async readFont(path)
//
// readFont takes a path which may be a url, and is async so needs await.

const unYetGul = await readFont('https://jsxcad.js.org/ttf/UnYetgul.ttf');

// #### Text(font, text, size=10)
// Text takes a font, the text to render, and an optional size defaulting to 10.

Text(unYetGul, '33').by(align('xy')).gridView();
```

**Reply by pentacular** [reply]
Yes, I think it needs a view mode in addition to edit mode.

The problem with comments is that you can't produce them using code.
